lib/taurus/qt/qtgui/taurusgui/config_loader/sardana.py

```diff
@@ -0,0 +1,206 @@
+#############################################################################
+##
+# This file is part of Taurus
+##
+# http://taurus-scada.org
+##
+# Copyright 2011 CELLS / ALBA Synchrotron, Bellaterra, Spain
+##
+# Taurus is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+##
+# Taurus is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+##
+# You should have received a copy of the GNU Lesser General Public License
+# along with Taurus.  If not, see <http://www.gnu.org/licenses/>.
+##
+###########################################################################
+
+""""""
+
+import sys
+
+import taurus
+from taurus.external.qt import Qt
+from taurus.qt.qtgui.taurusgui.config_loader.abstract import \
+    AbstractConfigLoader
+from taurus.qt.qtgui.taurusgui.utils import PanelDescription
+
+__all__ = ["SardanaConfigLoader"]
+
+
+class SardanaConfigLoader(AbstractConfigLoader):
+    """
+    Config loader which loads Sardana-related values.
+    This lives as a hack for manipulating AbstractLoader.CONFIG_VALUES.
+    It does not load anything by itself, just injects new values into
+    CONFIG_VALUES so other loaders will load them.
+    """
+
+    SARDANA_VALUES = [
+        "MACROSERVER_NAME",
+        "MACRO_PANELS",
+        "DOOR_NAME",
+        "MACROEDITORS_PATH",
+        "INSTRUMENTS_FROM_POOL",
+    ]
+
+    def __init__(self, confname):
+        super(SardanaConfigLoader, self).__init__(confname)
+        AbstractConfigLoader.CONFIG_VALUES + self.SARDANA_VALUES
```

Comment:
I don't know how safe is to monkeypatch the **class member**  AbstractConfigLoader.CONFIG_VALUES.

I have the impression that this may cause problems with other loaders which may not be ready to handle the injected keys.

Why not create set of less-hacky Sardana loaders? (SardanaPyConfigLoader, SardanaXmlConfigLoader, etc?) The hooks and most of the logic can be implemented in a base class that they all inherit.

Reply:
I admit it is very hacky but that was the simplest solution I could think of. I admit, it may not be super safe or even safe at all. From my point of view, it could work. But okay, I'll try to unhackify it.

Reply:
I've reimplemented it in (I hope so) safe way for py and xml. I could do this for json also but then I got to thinking: if someone will create new loader plugin (YAML/TOML), loading Sardana stuff will not work out of box. One would need to introduce extra loader (very simple on, but still) just in order to load also Sardana stuff.
Previous would work for all loaders out of the box. I wonder if we can have it both ways.

Reply:
After thinking about it for a while, one of solutions I can think of is using module variable as singleton. It would keep a list of fields to extract from config source. Only problem is that config loader has to be generic enough to support extracting data of different formats simply by referencing its name. JSON and Python ones are able to do it. I believe that some others like YAML or TOML would support it also. I do not know that would be exactly implemented when it comes to other types like databases.

Problem occurs with XML - I am by no means XML expert and now it looks like extracting some complex types like lists of dictionaries is really PITA. Ha, extracting even simple ones but with *type other than string* seams to be challenging. Only think I can think of is to count on some XML-experienced Python developer help to get it right and generic enough. And implement injecting extra config values into `CONFIG_VALUES` list, be it implemented as base class variable, module's variable or anything else. 

I think now two kinds of APIs are exposed by `ConfigLoader` - one from extracting data from source and one for mutating TaurusGUI (hooks). At first sight it looks bad (`do one thing and do it right`) but after a while, I think this is the right place to do it. It would be nice to have mutating API capability to not only add but also delete or even completely override list of loaded values from configuration. That way, whole universe of different configuration options is possible. The problem with that approach is it can be not that simple - there might be some mechanism that requires changes in many places to be implemented (vide synoptic and instrument panels - all over the code). Honestly, I do know how to approach it and any help in case of architectural decisions is appreciated.